src/ring_theory/henselian.lean

```diff
@@ -0,0 +1,189 @@
+    refine ih.add _,
+    rw [smodeq.zero, ideal.neg_mem_iff],
+    refine I.mul_mem_right _ _,
+    rw [← smodeq.zero] at h₁ ⊢,
+    exact (ih.eval f).trans h₁, },
+  have hfc : ∀ n, is_unit (f'.eval (c n)),
+  { intro n, contrapose! h₂,
+    rw [← mem_nonunits_iff, ← local_ring.mem_maximal_ideal, ← smodeq.zero] at h₂ ⊢,
+    exact ((hc' n).symm.eval _).trans h₂, },
+  have Hc : ∀ n, f.eval (c n) ∈ I ^ (n+1),
+  { intro n,
+    induction n with n ih, { simpa only [pow_one] },
+    simp only [nat.succ_eq_add_one],
+    rw [← taylor_eval_sub (c n), hc],
+    simp only [sub_eq_add_neg, add_neg_cancel_comm],
+    rw [eval_eq_sum, sum_over_range' _ _ _ (lt_add_of_pos_right _ zero_lt_two),
+      ← finset.sum_range_add_sum_Ico _ (nat.le_add_left _ _)],
+    swap, { intro i, rw zero_mul },
+    refine ideal.add_mem _ _ _,
+    { simp only [finset.sum_range_succ, taylor_coeff_one, mul_one, pow_one, mul_neg_eq_neg_mul_symm,
+        taylor_coeff_zero, finset.sum_singleton, finset.range_one, pow_zero],
+      rw [mul_left_comm, hinv _ (hfc n), mul_one, add_neg_self],
+      exact ideal.zero_mem _ },
+    { refine submodule.sum_mem _ _, simp only [finset.Ico.mem],
+      rintro i ⟨h2i, hi⟩,
+      have aux : n + 2 ≤ i * (n + 1), { transitivity 2 * (n + 1); nlinarith only [h2i], },
+      refine ideal.mul_mem_left _ _ (ideal.pow_le_pow aux _),
+      rw [pow_mul'],
+      refine ideal.pow_mem_pow ((ideal.neg_mem_iff _).2 $ ideal.mul_mem_right _ _ ih) _, } },
+  -- we are now in the position to show that `c : ℕ → R` is a Cauchy sequence
+  have aux : ∀ m n, m ≤ n → c m ≡ c n [SMOD (I ^ m • ⊤ : ideal R)],
+  { intros m n hmn,
+    rw [← ideal.one_eq_top, algebra.id.smul_eq_mul, mul_one],
+    obtain ⟨k, rfl⟩ := nat.exists_eq_add_of_le hmn, clear hmn,
+    induction k with k ih, { rw add_zero, },
+    rw [nat.succ_eq_add_one, ← add_assoc, hc, ← add_zero (c m), sub_eq_add_neg],
+    refine ih.add _, symmetry,
+    rw [smodeq.zero, ideal.neg_mem_iff],
+    refine ideal.mul_mem_right _ _ (ideal.pow_le_pow _ (Hc _)),
+    rw [add_assoc], exact le_self_add },
+  -- hence the sequence converges to some limit point `a`, which is the `a` we are looking for
+  obtain ⟨a, ha⟩ := is_precomplete.prec' c aux,
+  refine ⟨a, _, _⟩,
+  { show f.is_root a,
+    suffices : ∀ n, f.eval a ≡ 0 [SMOD (I ^ n • ⊤ : ideal R)], { from is_Hausdorff.haus' _ this },
+    intro n, specialize ha n,
+    rw [← ideal.one_eq_top, algebra.id.smul_eq_mul, mul_one] at ha ⊢,
+    refine (ha.symm.eval f).trans _,
+    rw [smodeq.zero],
+    exact ideal.pow_le_pow le_self_add (Hc _), },
+  { show a - a₀ ∈ maximal_ideal R,
+    specialize ha 1,
+    rw [hc, pow_one, ← ideal.one_eq_top, algebra.id.smul_eq_mul, mul_one, sub_eq_add_neg] at ha,
+    rw [← smodeq.sub_mem, ← add_zero a₀],
+    refine ha.symm.trans (smodeq.refl.add _),
+    rw [smodeq.zero, ideal.neg_mem_iff],
+    exact ideal.mul_mem_right _ _ h₁, }
+end
+
+/-
+TODO: After a good API for etale ring homomorphisms has been developed,
```

Comment:
Why not put this in the module docstring?

Reply:
Done